File: frontend/plugins/matching/patent.js
const matchesFilter = (patent, { primary, secondary, terciary }) => {
  let primaryMatch = true;
  let secondaryMatch = true;
  let terciaryMatch = true;

  if (primary.length != 0) {
    primaryMatch =
      primary.includes(patent.classification.primary.cip.substr(0, 1)) ||
      primary.includes(patent.classification.secondary?.cip.substr(0, 1));
  }

  if (primaryMatch && secondary.length != 0) {
    secondaryMatch =
      secondary.includes(patent.classification.primary.subarea) ||
      secondary.includes(patent.classification.secondary?.subarea);
  }
  const status = terciary[0];
  if (primaryMatch && secondaryMatch && status)
    terciaryMatch = status === patent.status;

  return primaryMatch && secondaryMatch && terciaryMatch;
};

export default (_, inject) => {
  inject("patentMatchesFilter", matchesFilter);
};

File: server/catalog/src/main/kotlin/br/usp/inovacao/hubusp/server/catalog/Patent.kt
package br.usp.inovacao.hubusp.server.catalog

@kotlinx.serialization.Serializable
data class Patent(
    val classification: DualClassification,
    val countries_with_protection: Set<String>,
    val inventors: Set<String>,
    val ipcs: Set<String>,
    val name: String,
    val owners: Set<String>,
    val photo: String? = null,
    val status: String,
    val summary: String,
    val url: String? = null
)

@kotlinx.serialization.Serializable
data class DualClassification(
    val primary: Area,
    val secondary: Area? = null
)

@kotlinx.serialization.Serializable
data class Area(
    val cip: String,
    val subarea: String
)
